src/internal/group_api/mod.rs

```diff
@@ -359,88 +372,153 @@ pub(crate) fn get_group_keys<'a>(
-                        let encrypted_group_key = recrypt.encrypt(
-                            &plaintext,
-                            &user_master_pub_key.into(),
-                            &auth.signing_private_key().into(),
-                        )?;
-                        // Map from UserId to (PublicKey, Optional TransformKey)
-                        // TransformKey is optional because we need it for members, but won't need it
-                        // when we expand this to include admins
-                        let member_info_result: Result<
-                            HashMap<UserId, (PublicKey, Option<TransformKey>)>,
-                            _,
-                        > = member_ids_and_keys
-                            .into_iter()
-                            .map(|(id, user_pub_key)| {
-                                let maybe_transform_key = recrypt.generate_transform_key(
-                                    &group_priv_key.clone().into(),
-                                    &user_pub_key.clone().into(),
-                                    &auth.signing_private_key().into(),
-                                );
-                                maybe_transform_key.map(|transform_key| {
-                                    (id, (user_pub_key, Some(transform_key.into())))
-                                })
+    use futures::future::Either;
+    let owner_id = match owner.clone() {
+        Some(id) => id,
+        None => auth.account_id().clone(),
+    };
+    // don't make any requests if validation fails
+    if admins.is_empty() {
+        Either::A(err(IronOxideErr::ValidationError(
+            format!("admins"),
+            format!("admins list cannot be empty"),
+        )))
+    } else if !admins.contains(&owner_id) {
+        Either::A(err(IronOxideErr::ValidationError(
+            format!("admins"),
+            format!("admins list must contain the owner"),
+        )))
+    } else {
+        Either::B(
+            user_api::user_key_list(auth, users_to_lookup)
+                .and_then(move |user_ids_and_keys| {
+                    // this will occur when one of the UserIds cannot be found
+                    if user_ids_and_keys.len() != users_to_lookup.len() {
+                        // figure out which user ids could not be found in the database and include them in the error message.
+                        let desired_users_set: HashSet<UserId> =
+                            HashSet::from_iter(users_to_lookup.clone());
+                        let found_users_set: HashSet<UserId> =
+                            user_ids_and_keys.into_iter().map(|(x, _)| x).collect();
+                        let diff: HashSet<&UserId> =
+                            desired_users_set.difference(&found_users_set).collect();
+                        err(IronOxideErr::UserDoesNotExist(format!(
+                            "Failed to find the following users: {:?}",
```

Comment:
`{}` vs `{:?}` One prints the debug version of the struct and one prints the display version.

Reply:
`{}` is Display whereas `{:?}` is Debug. We probably don't ever want to use Debug in error messages.

Reply:
Using `{}` fails to compile: ````std::collections::HashSet<&internal::user_api::UserId>` doesn't implement `std::fmt::Display````.

Using `{:?}` gives the IronOxideErr: `Error: UserDoesNotExist("Failed to find the following users: {UserId(\"aretheseuserids\"), UserId(\"likeafox\"), UserId(\"whatsthenextwhat\")}")`.

Is that not what we'd want it to display? If so, how do I implement Display for a HashSet of &UserId?

Reply:
I don't think any of the `std::collections` implement Display. This might be helpful https://doc.rust-lang.org/rust-by-example/hello/print/print_display.html